keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,740 @@
+  - Protect `ConfigMap` while being used by `Pod`
+  - Protect `StorageClass` while being used by `PersistentVolume`
+  - Protect `VolumeSnapshot` while being used by `VolumeSnapshotContent`
+  - Protect `VolumeSnapshotClass` while being used by `VolumeSnapshotContent`
+  - Protect `User`, `Group`, `ServiceAccount` and `Role` from `RoleBinding`
+  - Protect `User`, `Group`, `ServiceAccount` and `ClusterRole` from `ClusterRoleBinding`
+  - Protect dependent resources while owner resources aren't request to be deleted
+- To avoid accidental deletion of important resources:
+  - Protect resources that controllers or users marked as important
+
+### Goals
+
+- Provide a generic mechanism that prevents a resource, including CRD, from being deleted when it shouldn't be deleted
+- Implement the feature as an advisory feature
+
+### Non-Goals
+
+- Provide a specific mechanism to decide which particular objects should be prevented from being deleted when other particular objects exist
+- Provide a generic mechanism that prevents a resource from being updated when it shouldn't be updated
+- Implement any of the above potential use cases
+
+## Proposal
+
+A new field `Liens` to mark the object not to be deleted is introduced in object Metadata.
+Deletion requests for resources with non-empty `Liens` will be blocked by a newly introduced validation in api-server.
+
+## Glossary
+
+- Lien: An indication that some entity is relying on an API object. Deletion of objects with liens will be prevented until the interested party releases the lien.
+
+### User Stories (Optional)
+
+#### Story 1
+
+Another controller to protect Secret from deletion watches all Secrets and their potential user objects, like `Pod`, `PersistentVolumes`, and `VolumeSnapshotContent`. Once it find the Secret is used by one of the objects, it updates the Secret's `Liens` and ask this mechanism to block deletion request for the secret.
+
+#### Story 2
+
+A user knows this is an important object, and that terraform likes to delete and recreate objects, so the user sets a lien so that an accidental deletion will fail.
+
+<!--
+### Notes/Constraints/Caveats (Optional)
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+Most of the design ideas come from [here](https://github.com/kubernetes/kubernetes/issues/10179).
+Users or controllers can add a string to `Liens` to ask to protect a resource.
+A newly introduced validation in api-server will reject deletion requests for resources with non-empty `Liens` and return ["409 Conflict" error code](https://datatracker.ietf.org/doc/html/rfc2616#section-10.4.10).
```

Comment:
This puts the onus on clients to keep retrying if they want something to be deleted, right?

I prefer the declarative, eventually consistent nature of the current PVC protection implementation where a client can declare that it wants a thing to be deleted, and once it's safe to delete the thing, it is. We've built similar machinery to this in Crossplane and I believe it works pretty well.

For our use case "eventually consistent deletes" has nice symmetry with other operations. Most Crossplane resources correspond to some external system like a cloud provider. Let's say I want to create a GCP Subnet with Crossplane, but doing so requires the VPC it lives in to exist first. We don't reject the creation of the Subnet object in the API server because the VPC doesn't exist. Instead we accept the desired state (that a Subnet should exist) and our controller keeps trying to create it. Presumably it will eventually succeed when when the VPC exists.

Is there some way to address the requirement that deletes "don't start" (in the controller-reconciling-the-delete) sense without out-right rejecting the desire to delete a resource by an entity that has permission to do so?

Reply:
> where a client can declare that it wants a thing to be deleted, and once it's safe to delete the thing, it is

This is a very different semantic than "I want to delete this" -> "no".  The only way I know to achieve what you describe is with a finalizer, but this is not the right semantic.  At that point, the object in question is VISIBLY being deleted and async controllers are doing their own cleanup.  This is way to signal "begin the end-of-life process", not a way to delay it it.

Is there a different mechanism you are thinking of?

Reply:
> This is a very different semantic than "I want to delete this" -> "no".

Right. This KEP appears to propose that all forms of "in-use protection" use "I want to delete this" -> "no" semantics. Those semantics appear to be different from how the contemporary PV and PVC implementations of in-use protection work, which are more like "I want to delete this" -> "okay, eventually". For Crossplane's use-case, I prefer the semantics of the contemporary implementations work to what is proposed by this KEP.

My understanding is that PV and PVC in-use protection are more like "I want to delete this" -> "okay, eventually, when not in-use". The deleted objects transition into a `Terminating` state and go away once they're no longer in-use. We have the same "okay, eventually" semantics for Crossplane's `ProviderConfig` in-use protection, and are looking to generalize the pattern.

> Is there a different mechanism you are thinking of?

I'm not sure what you're asking. Can you elaborate?

To be clear - I think I understand the issues with finalizers and agree that they're not a good solution. I also don't have a good solution in mind. 🙂 What I would like is that _somehow_:

* If a client asks for an object be deleted
* The object can't be deleted _yet_ because doing so would break some other, dependent object
* The delete (i.e. desired state) be accepted and reconciled when possible

I don't think my preferred semantics necessarily make sense for all forms of deletion protection. The "Motivation" section of this KEP states that it's attempting to solve for two use cases:

1. Guaranteeing proper order of deletion (i.e. the PV and PVC use case, and Crossplane's)
2. Avoiding accidental deletion of important resources

I feel that "I want to delete this" -> "no" semantics makes sense for the second use case, but not the first.

Reply:
In KubeVirt we have a use-case with same or similar requirements:
1. We are running our processes in the original pod
2. There is an eviction request for this pod
3. Upon this request we want to transfer some state of the processes to a new pod
4. Upon the completed state transfer, the original pod is now free to be terminated

Important aspects are
a. Application can observe eviction request
b. Pod stays running until the state is transfered
c. caller of eviction is aware that the pod is preparing a shut-down

Reply:
@fabiand We could model this a few ways, but evictions are generally "urgent", right?  It's not clear to me how that should intersect with any sort of deletion delay.

Reply:
@thockin bringing your [related comment](https://github.com/kubernetes/enhancements/pull/2840#issuecomment-1585307886) back into this thread: 😄 

> To do what you want, we need a new lifecycle which includes some indefinite pre-delete state.

That may be the case, but what we want does seem in-scope for the goals defined by this design - i.e. guaranteeing proper order of deletion.

Take this manifest:

<details>
<summary> pod-with-volume.yaml </summary>

```yaml
---
apiVersion: v1
kind: Pod
metadata:
  name: task-pv-pod
spec:
  volumes:
    - name: task-pv-storage
      persistentVolumeClaim:
        claimName: task-pv-claim
  containers:
    - name: task-pv-container
      image: nginx
      ports:
        - containerPort: 80
          name: "http-server"
      volumeMounts:
        - mountPath: "/usr/share/nginx/html"
          name: task-pv-storage
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: task-pv-claim
spec:
  storageClassName: manual
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 3Gi
---
apiVersion: v1
kind: PersistentVolume
metadata:
  name: task-pv-volume
  labels:
    type: local
spec:
  storageClassName: manual
  capacity:
    storage: 10Gi
  accessModes:
    - ReadWriteOnce
  hostPath:
    path: "/mnt/data"
```

</details>

Today I can `kubectl apply -f pod-with-volume.yaml` and despite there being dependencies between the resources, the create just works and the controllers eventually arrive at my desired state - that they're created. I can `kubectl delete -f pod-with-volume.yaml` too. My desired state (that all this stuff be deleted) is immediately accepted, and eventually it's all gone.

I believe under this proposal the `kubectl delete -f pod-with-volume.yaml` UX would change. The `Pod` would delete immediately but the `PersistentVolume` and `PersistentVolumeClaim` would fail with a 409 Conflict, and I'd basically have to retry a couple of times for everything to actually get deleted.

This does technically achieve the goal of guaranteeing proper order of deletion - it just seems like a pretty poor UX for doing so.

Reply:
@thockin I was not ware that evictions had the notion of urgency.
Evictions respect PDBs, evictions are less hard than DELETE. Thus isn't eviction rather the less urgent approach and delete the real urgent way of getting rid of workloads?
fwiw `kubectl drain --force --grace-period=0` is urgent.

Thus to me a delay to an eviction is similar to [systemd inhibitors](https://www.freedesktop.org/software/systemd/man/systemd-inhibit.html) - a way to allow workloads to shut down gracefully. or even fail. And that's the difference: Fail gracefully, give admin/owner time to eeact to this. Before an admin jumps in with `--force --grace-period=0`